### ISSUE TITLE
OAS 3 Document has examples in the wrong path

### ISSUE THREAD

**Issue by d11wtq** [rswag/rswag#325]
According to the README:

> **Enable auto generation examples from responses**
> To enable examples generation from responses add callback above run_test! like:

```
after do |example|
  example.metadata[:response][:examples] = { 'application/json' => JSON.parse(response.body, symbolize_names: true) }
end
```

However this results in the examples being directly on the `response` object, whereas it should be on the media type object of the response's `content` property.

```
Structural error at paths./api/v1/accounts.get.responses.200
should NOT have additional properties
additionalProperty: examples
Jump to line 158
```

Any suggestions for a way to work around this?

**Comment by olegykz** [comment]
As a quick workaround, I had to monkey patch the formatter:
```ruby
module Rswag
  module Specs
    class SwaggerFormatter < ::RSpec::Core::Formatters::BaseTextFormatter

      def upgrade_content!(mime_list, target_node)
        target_node[:content] ||= {} # Here we're avoiding "content" key overriding
        schema = target_node[:schema]
        return if mime_list.empty? || schema.nil?

        mime_list.each do |mime_type|
          # TODO upgrade to have content-type specific schema
          (target_node[:content][mime_type] ||= {}).merge!(schema: schema)
        end
      end

    end
  end
end
```

And the examples collecting should look like
```ruby
  after do |example|
    example.metadata[:response][:content] = {
      'application/json' => { example: ... }
    }
  end
```

**Comment by smeevil** [comment]
Hi @olegykz,

I can confirm that your workaround works as advertised :)

Gerard.

**Comment by akofink** [comment]
This also happens for hard-coded response examples using the `example` DSL keyword.

**Comment by bob** [comment]
Hello @olegykz,
How did you patch a private `upgrade_content!` method? 
I'm trying to do it in `lib` directory and no chance, even via inheritance the `Rswag::Specs::SwaggerFormatter`.

**Comment by olegykz** [comment]
Hey @bob,
I'm not sure what's the difference between patching private/public/protected methods. Have you checked that your file in `lib` directory is actually loaded?

**Comment by chunlea** [comment]
> Hello @olegykz,
> How did you patch a private `upgrade_content!` method?
> I'm trying to do it in `lib` directory and no chance, even via inheritance the `Rswag::Specs::SwaggerFormatter`.

I also didn't found the correct way to overwrite this method, so I did this in `swagger_helper.rb`:

```ruby
  config.before(:all) do
    # A quick workaround unless rswag fix it.
    # https://github.com/rswag/rswag/issues/325#issuecomment-638695823
    module Rswag
      module Specs
        class SwaggerFormatter < ::RSpec::Core::Formatters::BaseTextFormatter

          def upgrade_content!(mime_list, target_node)
            target_node[:content] ||= {} # Here we're avoiding "content" key overriding
            schema = target_node[:schema]
            return if mime_list.empty? || schema.nil?

            mime_list.each do |mime_type|
              # TODO upgrade to have content-type specific schema
              (target_node[:content][mime_type] ||= {}).merge!(schema: schema)
            end
          end
        end
      end
    end
  end
```

**Comment by Bandes** [comment]
I haven't been able to get this workaround to work for my examples. Has there been any progress on the PR that would fix this?

**Comment by rdnewman** [comment]
I wanted automatic generation of both multiple response AND payload examples.  At the point, it got a little thick, so threw auto-generation of path parameters too.

Can see solution at https://gist.github.com/rdnewman/ba5143f1598eff72b2cba48ce78ca43d.  Most files should be plug and play (but watch the top level module names). Only tested with my own project that I did this for (using Redoc), so YMMV.

**Comment by phlegx** [comment]
The workaround don't work for me with Rswag version 2.3.1. I have tested both workarounds described here. What do I wrong? Can you show an example with commands please?

**Found the problem**

Running `rake rswag` or `rails rswag` with `RAILS_ENV=test` the `swagger_dry_run` is ignored!

```ruby
# config/environments/test.rb

RSpec.configure do |config|
  config.swagger_dry_run = false
end
```

See issue #359!

**Comment by M-H-Mukit** [comment]
@phlegx  try putting the code in `config/application.rb` instead 

`swagger_dry_run` issue will be solved.

After that, I stuck on. The example is not auto-generating in swagger docs.

Please let me know if you can figure out.

**Comment by phlegx** [comment]
@M-H-Mukit putting code at the end of file config/application.rb works! Thx

**Comment by phlegx** [comment]
**Rswag:** 2.3.1
**Rails:** 6.0.3.x

Here my code how to fix in the meantime the examples problem:

```ruby
...
  # spec/swagger_helper.rb
  # A quick workaround unless rswag fix it.
  # upgrade_content! - https://github.com/rswag/rswag/issues/325
  # stop - https://github.com/rswag/rswag/pull/304
  config.before(:all) do
    module Rswag
      module Specs
        # Hook SwaggerFormatter class method.
        class SwaggerFormatter < ::RSpec::Core::Formatters::BaseTextFormatter
          ::RSpec::Core::Formatters.register(self, :example_group_finished, :stop)

          def upgrade_content!(mime_list, target_node)
            # Fix examples with the following line change.
            target_node[:content] ||= {}
            schema = target_node[:schema]
            return if mime_list.empty? || schema.nil?

            mime_list.each do |mime_type|
              # Fix examples with the following line change.
              (target_node[:content][mime_type] ||= {}).merge!(schema: schema)
            end
          end

          def stop(_notification = nil)
            @config.swagger_docs.each do |url_path, doc|
              unless doc_version(doc).start_with?('2')
                doc[:paths]&.each_pair do |_k, v|
                  v.each_pair do |_verb, value|
                    is_hash = value.is_a?(Hash)
                    if is_hash && value.dig(:parameters)
                      schema_param = value.dig(:parameters)&.find { |p| (p[:in] == :body || p[:in] == :formData) && p[:schema] }
                      mime_list = value.dig(:consumes)
                      if value && schema_param && mime_list
                        value[:requestBody] = { content: {} } unless value.dig(:requestBody, :content)
                        mime_list.each do |mime|
                          value[:requestBody][:content][mime] = { schema: schema_param[:schema] }
                          # Fix examples with the following line.
                          value[:requestBody][:content][mime].merge!(examples: schema_param[:examples]) if schema_param[:examples]
                        end
                      end

                      value[:parameters].reject! { |p| p[:in] == :body || p[:in] == :formData }
                    end
                    remove_invalid_operation_keys!(value)
                  end
                end
              end

              file_path = File.join(@config.swagger_root, url_path)
              dirname = File.dirname(file_path)
              FileUtils.mkdir_p dirname unless File.exist?(dirname)

              File.open(file_path, 'w') do |file|
                file.write(pretty_generate(doc))
              end

              @output.puts "Swagger doc generated at #{file_path}"
            end
          end
        end
      end
    end
  end
```
And the examples collecting should look like:

```ruby
# spec/requests/...

after do |example|
  example.metadata[:response][:content] = {
      'application/json' => { example: JSON.parse(response.body, symbolize_names: true) }
  }
end
```

**Comment by M-H-Mukit** [comment]
@phlegx  Thank you so much. Will give it a try

Update: It works! Thank you again.

**Comment by matiasgarcia** [comment]
@phlegx Can you walk me through how you fixed it?

I tried adding `config.swagger_dry_run = false` to both `environments/test.rb` and `application.rb`.
Then I added your monkey patch on `swagger_helper.rb`, which worked.

Then, at the top of the spec I added the after hook:

```
describe 'Withdrawals API', swagger_doc: 'v1/swagger.yaml' do
  path '/api/v1/withdrawals' do
    after do |example|
      example.metadata[:response][:content] = {
        'application/json' => JSON.parse(response.body, symbolize_names: true)
      }
    end
```

Which seems to be running.

Then when I run `rails rswag` this is the output:

```
mgarcia@BTX-NTB-030:~/projects/withdrawals$ rails rswag
/home/mgarcia/.rbenv/versions/2.5.7/bin/ruby -I/home/mgarcia/.rbenv/versions/2.5.7/lib/ruby/gems/2.5.0/gems/rspec-core-3.10.0/lib:/home/mgarcia/.rbenv/versions/2.5.7/lib/ruby/gems/2.5.0/gems/rspec-support-3.10.0/lib /home/mgarcia/.rbenv/versions/2.5.7/lib/ruby/gems/2.5.0/gems/rspec-core-3.10.0/exe/rspec --pattern spec/requests/\*\*/\*_spec.rb,\ spec/api/\*\*/\*_spec.rb,\ spec/integration/\*\*/\*_spec.rb --format Rswag::Specs::SwaggerFormatter --dry-run --order defined
Generating Swagger docs ...
Swagger doc generated at /home/mgarcia/projects/withdrawals/swagger/v1/swagger.yaml

Finished in 0.0013 seconds (files took 2.18 seconds to load)
5 examples, 0 failures

Coverage report generated for RSpec to /home/mgarcia/projects/withdrawals/coverage. 61 / 337 LOC (18.1%) covered.
```

As you see it's doing `--dry-run`, so I tried doing:

```
mgarcia@BTX-NTB-030:~/projects/withdrawals$ bundle exec rspec spec/requests/ --format Rswag::Specs::SwaggerFormatter
Generating Swagger docs ...
Swagger doc generated at /home/mgarcia/projects/withdrawals/swagger/v1/swagger.yaml

Finished in 0.18486 seconds (files took 2.18 seconds to load)
5 examples, 0 failures

Coverage report generated for RSpec to /home/mgarcia/projects/withdrawals/coverage. 105 / 314 LOC (33.44%) covered.
```

But what I get is still generic examples and not the ones that the spec returns:

![Screenshot from 2020-12-29 12-27-41](https://user-images.githubusercontent.com/2261869/103294800-4e903c00-49d1-11eb-9c66-9cc20eeb3cbc.png)

This can be easily seen because it returns "id: 0" when it should be a value `> 0`

An example of the `swagger.yaml`:

```
  "/api/v1/withdrawals":
    post:
      parameters: []
      responses:
        '200':
          description: Creates or returns an already generated Withdrawal with the
            given key
          content:
            application/json:
              id: 50
              status: pending_submit
              key: unique_key
              schema:
                "$ref": "#/components/schemas/Withdrawal"
```

Running:

rspec-rails 4.0.1
rspec 3.10.0
rswag (2.3.1)
rails (6.1.0)

I filled the app of logs and the overrides and the after hooks are being called. So I am running in circles at this point.

**Comment by helderasmoreira** [comment]
I'm wondering why this was closed. Has this been fixed and present in a recent release?

**Comment by oblakeerickson** [comment]
@heldersantosmoreira yep it was automatically closed as part of merging in changes from #371. Checkout the latest release 2.4.0.

**Comment by helderasmoreira** [comment]
@oblakeerickson you're right, was on 2.3.2. Thanks for letting me know! 🙏
